src/kde_analysis/eval_m1m2chieff_poiss_weights.py

```diff
@@ -103,18 +102,274 @@ def get_rate_m_chieff2D(m1_query, m2_query, Rate):
+
+    # Validate keep_dims
+    if np.any(keep_dims < 0) or np.any(keep_dims >= n_dims):
+        raise ValueError(f"keep_dims must be in range [0, {n_dims-1}]")
+
+    kept_names = [dimension_names[i] for i in keep_dims]
+    removed_names = [dimension_names[i] for i in range(n_dims) if i not in keep_dims]
+
+    # 1. Marginalize data (just select columns)
+    data_marginalized = data_nd[:, keep_dims]
+
+    # 2. Marginalize transformations
+    if input_transf is not None:
+        input_transf_marg = tuple(input_transf[i] for i in keep_dims)
+    else:
+        input_transf_marg = None
+
+    # 3. Marginalize rescale factors
+    if rescale_factors is not None:
+        rescale_marg = [rescale_factors[i] for i in keep_dims]
+    else:
+        rescale_marg = None
+
+    return {
+        'data': data_marginalized,
+        'bandwidth': per_point_bw,  
+        'input_transf': input_transf_marg,
+        'rescale': rescale_marg,
+        'weights': weights,
+        'kept_dims': list(keep_dims),
+        'dimension_names': kept_names,
+        'original_dims': n_dims
+    }
+
+
+def create_marginalized_kde(
+    samples,
+    per_point_bw,
+    keep_dims,
+    input_transf,
+    rescale_factors,
+    weights,
+    m1grid,
+    m2grid,
+    cfgrid,
+    alpha,
+    group,
+    dimension_names=None
+):
+    """
+    Create and evaluate a marginalized KDE for specified dimensions.
+    Uses symmetrize_dims in KDE objects when both m1 and m2 are present.
+
+    Parameters
+    ----------
+    samples : np.ndarray, shape (n_samples, 3)
+        Original 3D samples
+    per_point_bw : np.ndarray, shape (n_samples,)
+        Scalar per-point bandwidth for each sample
+    keep_dims : list of int
+        Dimensions to keep (e.g., [0,1] for 2D, [0] for 1D, [0,1,2] for 3D)
+    input_transf : tuple
+        Input transformations for all dimensions
+    rescale_factors : list
+        Rescale factors for all dimensions
+    weights : np.ndarray or None
+        Sample weights
+    m1grid, m2grid, cfgrid : np.ndarray
+        Grid arrays for each dimension
+    alpha : float
+        Adaptive bandwidth parameter (only used if 'perpoint_bws' not in group)
+    group : h5py.Group
+        HDF5 group containing 'perpoint_bws' key
+    dimension_names : list, optional
+        Names of dimensions (default: ['m1', 'm2', 'chieff'])
+
+    Returns
+    -------
+    dict with keys:
+        'kde_values': Evaluated KDE on grid (shape depends on n_dims)
+        'grid_shape': Shape of the grid
+        'eval_samples': Flattened evaluation samples
+        'marginalized_result': Output from marginalize_kde_data
+        'kde_object': The trained KDE object
+    """
+
+    if dimension_names is None:
+        dimension_names = ['m1', 'm2', 'chieff']
+
+    # Grid mapping
+    grid_map = {0: m1grid, 1: m2grid, 2: cfgrid}
+
+    n_dims = len(keep_dims)
+
+    # Step 1: Marginalize the data 
+    result = marginalize_kde_data(
```

Comment:
```suggestion
    # Marginalize (reduce dimensionality of) the KDE input data 
    marg_data = marginalize_kde_data(
```
'Step 1' etc are not necessary - also I think the name 'result' is confusing as it's just an intermediate step